src/diffcalc/ub/crystal.py

```diff
@@ -3,15 +3,34 @@
 A module defining crystal lattice class and auxiliary methods for calculating
 crystal plane geometric properties.
 """
+from inspect import signature
 from math import acos, cos, degrees, pi, radians, sin, sqrt
-from typing import Any, Dict, List, Optional, Tuple
+from typing import Callable, Dict, List, Tuple
 
 import numpy as np
-from diffcalc.util import allnum, angle_between_vectors, zero_round
+from diffcalc.util import DiffcalcException, angle_between_vectors, zero_round
 from numpy.linalg import inv
+from typing_extensions import TypedDict
 
 
-class Crystal:
+class Crystal(TypedDict):
```

Comment:
As before, I think it's best if we chose one thing out dataclasses, typeddicts and pydantic models and stick with it. I appreciate that this may be easier said than done though!